apps/labs/posts/multi-dimensional-sparse-arrays.md

```diff
@@ -0,0 +1,225 @@
+
+By the end of this stage, I had become fairly confident that the entire COO codebase could be extended to n-dimensions.
+
+I then moved on to the second stage, which involved tackling the fundamental operations in linear algebra, `matmul`, `dot`, and `tensordot`.
+
+Most functions in SciPy's sparse module fall into one of two categories: they are either purely Python-based or they involve compiled code. Compiled code is introduced when a significant speed boost is essential. However, increasing the amount of compiled code also increases the size of binaries, which can lead to portability issues and larger memory footprints. To add on, another issue with compiled code is maintenance. Not many people want to jump down into C++! Fortunately, Python's numeric library, NumPy, has been optimised so heavily that it can at times deliver as good of a performance as compiled code, if not faster, while also avoiding the drawbacks of increased memory usage associated with binaries.
+
+Given these considerations, the implementation of methods in this stage required careful thought. In the next section, I will walk through the key decisions made during the implementation of the `matmul`, `dot`, and `tensordot` methods. But first, I will introduce a concept called broadcasting.
+
+### What is broadcasting?
+
+Broadcasting is a technique used to expand arrays of different shapes to a given shape, making them compatible for element-wise operations. This process starts from the rightmost dimension and works leftwards. If both arrays have matching sizes in a particular dimension, we move on to the next. If one of the dimensions has size 1, it is "stretched" or repeated to match the other.
+
+<p align="center">
+  <img
+    alt="Visual illustration of broadcasting."
+    src="/posts/multi-dimensional-sparse-arrays/broadcasting.png"
+    width="67%"
+  />
+</p>
+<i align="center">
+  Retrieved from NumPy Documentation.
+</i>
+
+NumPy provides a method `broadcast_to`, which allows a dense array to be broadcast to a specific shape.
+
+The first task in this phase was to implement a method that replicates this behavior for COO arrays. However, it’s important to note that whie NumPy generates a "view" of the original array, the method we introduced instantiated a new `coo_array` object whose `data` and `coords` attributes were generated by tiling and repeating the original array's `data` and `coords`.
```

Comment:
```suggestion
The first task in this phase was to implement a method that replicates this behavior for COO arrays. However, it’s important to note that while NumPy generates a "view" of the original array, the method we introduced instantiated a new `coo_array` object whose `data` and `coords` attributes were generated by tiling and repeating the original array's `data` and `coords`.
```

Reply:
Also it would be good to add a short note on why this matters - this concept is far into the weeds for the regular reader, so it may not be obvious why you are mentioning this. If you want, you can also add a link to https://numpy.org/doc/stable/user/basics.copies.html

Reply:
Done at 8bc30fb55294d61a0f9a7cf13c6316cbdc85c462!